File: akka-essentials/Chapter04/Calculator/src/main/java/Calculator.java
package akka.examples;

import akka.actor.TypedActor;
import akka.dispatch.Futures;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Option;
import scala.concurrent.Future;

public class Calculator implements CalculatorInt {
    private Integer counter = 0;

    //Non-blocking request response
    public Future<Integer> add(Integer first, Integer second) {
        return Futures.successful(first + second); // TypedActor.dispatcher());
    }

    //Non-blocking request response
    public Future<Integer> subtract(Integer first, Integer second) {
        return Futures.successful(first - second); // TypedActor.dispatcher());
    }

    //fire and forget
    public void incrementCount() {
        counter++;
    }

    //Blocking request response
    public Option<Integer> incrementAndReturn() {
        return Option.some(++counter);
    }

    private LoggingAdapter log =
        Logging.getLogger(TypedActor.context().system(), this);

    //Allows to tap into the Actor PreStart hook
    public void preStart() {
        log.info("Actor Started !");
    }

    //Allows to tap into the Actor PostStop hook
    public void postStop() {
        log.info("Actor Stopped !");
    }

}

File: akka-essentials/Chapter04/CalculatorScala/src/main/scala/Calculator.scala
package akka.examples

import akka.actor.ActorSystem//TypedActor
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}
import akka.dispatch.Futures
//import akka.event.Logging
//import akka.event.LogSource
import scala.concurrent.Future

object Calculator {
  def apply(): Behavior[Nothing] =
    Behaviors.setup[Nothing](context => new Calculator(context))
}

class Calculator(context: ActorContext[Nothing])
    extends AbstractBehavior[Nothing](context) with CalculatorInt {

  private var counter: Int = 0

  def add(first: Int, second: Int): Future[Int] =
    Futures.successful(first + second)

  def subtract(first: Int, second: Int): Future[Int] =
    Futures.successful(first - second)

  def incrementCount(): Unit = counter += 1

  def incrementAndReturn(): Option[Int] = {
    counter += 1
    Some(counter)
  }

  //private val log = Logging(ActorSystem("Calculator"), this)

  def preStart(): Unit =
    context.log.info ("Actor Started")

  def postStop(): Unit =
    context.log.info ("Actor Stopped")

  override def onMessage(msg: Nothing): Behavior[Nothing] = {
    // no need to handle any messages
    Behaviors.unhandled
  }

}
